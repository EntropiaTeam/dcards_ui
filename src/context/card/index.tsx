import React, {
  createContext, useContext, useReducer, FC
} from 'react';
import { CardActionsType, UseCardContext } from './types/CardTypes';
import CardReducer, { initialState } from './reducer/CardReducer';
import { CardActions } from './actions/CardActions';

// Create Context
export const CardContext = createContext(initialState);
export const CardActionsContext = createContext({} as CardActionsType);
export const useCardContext = (): UseCardContext => (
  { state: useContext(CardContext), actions: useContext(CardActionsContext) }
);

export const CardProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CardReducer, initialState);

  // Actions
  const actions: CardActionsType = {
    setCardToEdit: CardActions.setCardToEdit(dispatch)
  };

  return (
    <CardContext.Provider value={state}>
      <CardActionsContext.Provider value={actions}>
        {children}
      </CardActionsContext.Provider>
    </CardContext.Provider>
  );
};
