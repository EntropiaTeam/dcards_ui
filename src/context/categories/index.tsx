import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CategoriesActionsType, UseCategoriesContext } from './types/CategoriesTypes';
import CategoriesReducer, { initialState } from './reducer/CategoriesReducer';
import { CategoriesActions } from './actions/CategoriesActions';

// Create Context
export const CategoriesContext = createContext(initialState);
export const CategoriesActionsContext = createContext({} as CategoriesActionsType);
export const useCategoriesContext = (): UseCategoriesContext => ({
  state: useContext(CategoriesContext), actions: useContext(CategoriesActionsContext)
});

export const CategoriesProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CategoriesReducer, initialState);

  // Actions
  const actions: CategoriesActionsType = useMemo(() => ({
    setCategories: CategoriesActions.setCategories(dispatch)
  }), []);

  return (
    <CategoriesContext.Provider value={state}>
      <CategoriesActionsContext.Provider value={actions}>
        {children}
      </CategoriesActionsContext.Provider>
    </CategoriesContext.Provider>
  );
};
