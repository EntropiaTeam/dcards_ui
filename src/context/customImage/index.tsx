import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { CustomImageActionsType, UseCustomImageContext } from './types/CustomImageTypes';
import { CustomImageActions } from './actions/CustomImageActions';
import CustomImageReducer, { initialState } from './reducer/CustomImageReducer';

// Create Context
export const CustomImageContext = createContext(initialState);
export const CustomImageActionsContext = createContext({} as CustomImageActionsType);
export const useCustomImageContext = (): UseCustomImageContext => (
  { state: useContext(CustomImageContext), actions: useContext(CustomImageActionsContext) }
);

export const CustomImageProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(CustomImageReducer, initialState);
  // Actions
  const actions: CustomImageActionsType = useMemo(() => ({
    setErrors: CustomImageActions.setErrors(dispatch),
    setUserImage: CustomImageActions.setUserImage(dispatch),
    setCropper: CustomImageActions.setCropper(dispatch),
    setAgreement: CustomImageActions.setAgreement(dispatch),
    setAgreementTranslation: CustomImageActions.setAgreementTranslation(dispatch),
    setCroppedImage: CustomImageActions.setCroppedImageValue(dispatch),
    setCropperBackImageURL: CustomImageActions.setCropperBackImageValue(dispatch),
    setRotationDegree: CustomImageActions.setRotationDegree(dispatch),
    setZoomRatio: CustomImageActions.setZoomRatio(dispatch),
    setCroppedCanvasData: CustomImageActions.setCroppedCanvasData(dispatch),
    resetCustomImageState: CustomImageActions.resetCustomImageState(dispatch)
  }), [dispatch]);

  return (
    <CustomImageContext.Provider value={state}>
      <CustomImageActionsContext.Provider value={actions}>
        {children}
      </CustomImageActionsContext.Provider>
    </CustomImageContext.Provider>
  );
};
