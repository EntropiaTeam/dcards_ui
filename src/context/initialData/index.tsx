import React, {
  createContext, useContext, useReducer, useMemo, FC
} from 'react';
import { UseInitialDataContext, InitialDataActionsType } from './types/InitialDataTypes';
import InitialDataReducer, { initialState as initDatainitialState } from './reducer/InitialDataReducer';
import { InitialDataActions } from './actions/InitialDataActions';

import {
  useCategoriesContext,
  useCustomImageContext
} from '..';

// Create Context
export const InitialDataContext = createContext(initDatainitialState);
export const InitialDataActionsContext = createContext({} as InitialDataActionsType);
export const useInitialDataContext = (): UseInitialDataContext => ({
  state: useContext(InitialDataContext), actions: useContext(InitialDataActionsContext)
});

export const InitialDataProvider: FC = ({ children }) => {
  const { actions: { setCategories } } = useCategoriesContext();
  const { actions: { setAgreementTranslation } } = useCustomImageContext();
  const [initDatastate, dispatchInitData] = useReducer(InitialDataReducer, initDatainitialState);

  // Actions
  const actions: InitialDataActionsType = useMemo(() => ({
    loadInitialData: InitialDataActions.loadInitialData(
      dispatchInitData,
      setCategories,
      setAgreementTranslation
    )
  }), [setCategories, setAgreementTranslation]);

  return (
    <InitialDataContext.Provider value={initDatastate}>
      <InitialDataActionsContext.Provider value={actions}>
        {children}
      </InitialDataActionsContext.Provider>
    </InitialDataContext.Provider>
  );
};
