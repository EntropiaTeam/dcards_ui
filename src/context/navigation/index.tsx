import React, {
  createContext, useContext, useReducer, FC
} from 'react';
import { NavigationActionTypes, UseNavigationContext } from './types/NavigationTypes';
import NavigationReducer, { initialState } from './reducer/NavigationReducer';
import { NavigationActions } from './actions/NavigationActions';

// Create Context
export const NavigationContext = createContext(initialState);
const NavigationActionsContext = createContext({} as NavigationActionTypes);
export const useNavigationContext = (): UseNavigationContext => (
  { state: useContext(NavigationContext), actions: useContext(NavigationActionsContext) }
);

export const NavigationProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(NavigationReducer, initialState);

  // Actions
  const actions: NavigationActionTypes = {
    changeIsBackButtonVisible: NavigationActions.changeIsBackButtonVisible(dispatch),
    changeIframeParams: NavigationActions.changeIframeParams(dispatch)
  };

  return (
    <NavigationContext.Provider value={state}>
      <NavigationActionsContext.Provider value={actions}>
        {children}
      </NavigationActionsContext.Provider>
    </NavigationContext.Provider>
  );
};
