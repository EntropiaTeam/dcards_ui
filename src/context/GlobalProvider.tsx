import React from 'react';
import {
  CardProvider,
  CategoriesProvider,
  TextProvider,
  OrderProvider,
  CustomImageProvider
} from '.';
import { NavigationProvider } from './navigation';

export const GlobalProvider: React.FunctionComponent = ({ children }) => (
  <NavigationProvider>
    <CardProvider>
      <CategoriesProvider>
        <TextProvider>
          <CustomImageProvider>
            <OrderProvider>
              {children}
            </OrderProvider>
          </CustomImageProvider>
        </TextProvider>
      </CategoriesProvider>
    </CardProvider>
  </NavigationProvider>
);
