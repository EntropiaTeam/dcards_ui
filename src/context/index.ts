import { CardProvider, useCardContext } from './card';
import { CategoriesProvider, useCategoriesContext } from './categories';
import { TextProvider, useTextContext } from './text';
import { OrderProvider, useOrderContext } from './order';
import { CustomImageProvider, useCustomImageContext } from './customImage';
import { NavigationProvider, useNavigationContext } from './navigation';

export {
  CardProvider,
  useCardContext,
  CategoriesProvider,
  useCategoriesContext,
  TextProvider,
  useTextContext,
  OrderProvider,
  useOrderContext,
  CustomImageProvider,
  useCustomImageContext,
  NavigationProvider,
  useNavigationContext
};
