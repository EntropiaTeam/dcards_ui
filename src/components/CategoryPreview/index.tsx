import React, { FC } from 'react';
import {
  Grid, makeStyles, Theme, Typography
} from '@material-ui/core';
import { useHistory } from 'react-router-dom';
import { CardImage, ImageContainer } from '..';
import { useCardContext } from '../../context';
import { Category } from '../../context/categories/types/CategoriesTypes';
import useAppConfig from '../../hooks/useAppConfig';
import { getImageUrl } from '../../helpers';

type Props = {
  category: Category;
  tabIndex?: number;
};

const useStyles = makeStyles((theme: Theme) => ({
  categoryContainer: {
    display: 'flex',
    flexFlow: 'column',
    textAlign: 'center',
    margin: '27px 0 8px 0',
    cursor: 'pointer',
    '&:focus': {
      outline: 'none'
    },
    '&:focus-visible': {
      outline: 'auto',
      outlineColor: theme.palette.primary.main
    }
  },
  imagesContainer: {
    display: 'flex',
    justifyContent: 'center',
    paddingBottom: 16,
    paddingTop: 16
  },
  imgContainer: {
    border: '1px solid rgba(0, 0, 0, 0.1)',
    boxShadow: '0px 7px 49px rgba(202, 204, 213, 0.45)'
  }
}));

const CategoryPreview: FC<Props> = ({ category, tabIndex }) => {
  const classes = useStyles();
  const history = useHistory();
  const { appConfig } = useAppConfig();
  const { actions: { setCardToEdit } } = useCardContext();

  const imgConfigRotateToLeft = {
    ...appConfig.categoriesImgConfig,
    rotation: -8
  };

  const imgConfigRotateToRight = {
    ...appConfig.categoriesImgConfig,
    rotation: 8,
    leftMargin: -60
  };

  const handlePushToNextRoute = (): void => {
    if (appConfig.customCategoryId === category.id) {
      const card = category.cards[0];
      setCardToEdit(card);

      const targetLocation = `/categories/${category.id}/cards/${card.id}/edit/photo`;
      history.push({ pathname: `/categories/${category.id}/cards/agreement`, state: { targetLocation } });
    } else {
      history.push(`/categories/${category.id}/cards`);
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>): void => {
    if (event.key === 'Enter') {
      handlePushToNextRoute();
    }
  };

  const firstImgConfig = category.cards.length === 1
    ? appConfig.categoriesImgConfig
    : imgConfigRotateToLeft;

  const bottomPreview = category.preview_cards[0];
  const topPreview = category.preview_cards[1];
  return (
    <Grid
      item
      xs={6}
      sm={6}
      md={3}
      lg={3}
      xl={3}
      className={classes.categoryContainer}
      onClick={handlePushToNextRoute}
      onKeyDown={handleKeyDown}
      tabIndex={tabIndex}
    >
      <div className={classes.imagesContainer}>
        {bottomPreview && (
          <ImageContainer imgConfig={firstImgConfig} cls={classes.imgContainer}>
            <CardImage
              imageUrl={getImageUrl(`${bottomPreview}_front.png`, 'thumb')}
              cardName={bottomPreview}
            />
          </ImageContainer>
        )}
        {topPreview && (
        <ImageContainer imgConfig={imgConfigRotateToRight} cls={classes.imgContainer}>
          <CardImage
            imageUrl={getImageUrl(`${category.preview_cards[1]}_front.png`, 'thumb')}
            cardName={topPreview}
          />
        </ImageContainer>
        )}
      </div>
      <Typography>{category.name}</Typography>
    </Grid>
  );
};

export default CategoryPreview;
