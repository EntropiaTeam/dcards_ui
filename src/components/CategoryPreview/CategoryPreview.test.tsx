import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Grid } from '@material-ui/core';
import CategoryPreview from './index';
import { CardActionsContext } from '../../context/card';

interface ReactRouterDom {
  useHistory: () => void;
  useRouteMatch: () => void;
}

const category = {
  id: 'custom_photo',
  name: 'Custom Photo',
  culture_names: [
    { 'en-US': 'Custom Photo' },
    { 'en-CA': 'Custom Photo' },
    { 'fr-CA': 'Photo personnalisÃ©e' }
  ],
  occasion_maps: { 'en-US': [], 'en-CA': [], 'fr-CA': [] },
  preview_cards: ['blank_photo_portrait'],
  start_date: '',
  end_date: '',
  sequence: 40,
  cards: [
    {
      id: 'blank_photo_portrait',
      name: 'Full frame photo card',
      type: 'portrait',
      attributes: ['photo'],
      category: ['custom_photo', 'valentines_day', 'g_congrats_general', 'h_congrats_wedding'],
      tags: ['custom'],
      culture: ['en-US', 'en-CA'],
      font: {
        family: 'Roboto Slab',
        size: '14',
        weight: '500',
        color: '#000'
      }
    }
  ]
};

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', (): ReactRouterDom => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush
  }),
  useRouteMatch: jest.fn().mockReturnValue(true)
}));

describe('<CategoryPreview />', () => {
  const setCardToEdit = jest.fn(() => {});

  it('should render CategoryPreview component', () => {
    const categoryPreviewRenderer = renderer.create(
      <CategoryPreview category={category} />
    );
    expect(categoryPreviewRenderer.toJSON()).toMatchSnapshot();
  });

  it('should move from CategoryPreview to card page', () => {
    const wrapper = mount(
      <CardActionsContext.Provider value={{ setCardToEdit }}>
        <CategoryPreview category={category} />
      </CardActionsContext.Provider>
    );

    wrapper.find(Grid).simulate('click');
    expect(setCardToEdit).toHaveBeenCalledTimes(1);
    expect(setCardToEdit).toHaveBeenCalledWith(category.cards[0]);
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
    expect(mockHistoryPush).toHaveBeenCalledWith({
      pathname: '/categories/custom_photo/cards/agreement',
      state: { targetLocation: '/categories/custom_photo/cards/blank_photo_portrait/edit/photo' }
    });
  });
});
