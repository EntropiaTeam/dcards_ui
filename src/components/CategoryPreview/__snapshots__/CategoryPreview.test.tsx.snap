// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryPreview /> should render CategoryPreview component 1`] = `
<div
  className="MuiGrid-root makeStyles-categoryContainer-1 MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-md-3 MuiGrid-grid-lg-3 MuiGrid-grid-xl-3"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div
    className="makeStyles-imagesContainer-2"
  >
    <div
      className="MuiBox-root MuiBox-root-6 makeStyles-imageContainer-4 makeStyles-imageContainer-5 makeStyles-imgContainer-3"
      tabIndex={-1}
    >
      <img
        alt="blank_photo_portrait"
        className="makeStyles-image-7 makeStyles-image-8"
        src="/assets/i/blank_photo_portrait_thumb_front.png"
      />
    </div>
  </div>
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Custom Photo
  </p>
</div>
`;
